test/e2e/test/agent/builder.go

```diff
@@ -0,0 +1,309 @@
+
+	// if it exists, move PodTemplate from Deployment to DaemonSet
+	if b.Agent.Spec.Deployment != nil {
+		b.Agent.Spec.DaemonSet.PodTemplate = b.Agent.Spec.Deployment.PodTemplate
+		b.Agent.Spec.Deployment = nil
+	}
+
+	b.PodTemplate = &b.Agent.Spec.DaemonSet.PodTemplate
+
+	return b
+}
+
+func (b Builder) WithDeployment() Builder {
+	b.Agent.Spec.Deployment = &agentv1alpha1.DeploymentSpec{}
+
+	// if it exists, move PodTemplate from DaemonSet to Deployment
+	if b.Agent.Spec.DaemonSet != nil {
+		b.Agent.Spec.Deployment.PodTemplate = b.Agent.Spec.DaemonSet.PodTemplate
+		b.Agent.Spec.DaemonSet = nil
+	}
+	b.PodTemplate = &b.Agent.Spec.Deployment.PodTemplate
+
+	return b
+}
+
+func (b Builder) WithDeploymentStrategy(s appsv1.DeploymentStrategy) Builder {
+	modifiedBuilder := b
+	if b.Agent.Spec.Deployment == nil {
+		modifiedBuilder = b.WithDeployment()
+	}
+	modifiedBuilder.Agent.Spec.Deployment.Strategy = s
+	return modifiedBuilder
+}
+
+func (b Builder) WithDefaultESValidation(validation ValidationFunc) Builder {
+	return b.WithESValidation(validation, "default")
+}
+
+func (b Builder) WithESValidation(validation ValidationFunc, outputName string) Builder {
+	b.Validations = append(b.Validations, validation)
+	b.ValidationsOutputs = append(b.ValidationsOutputs, outputName)
+
+	return b
+}
+
+func (b Builder) WithElasticsearchRefs(refs ...agentv1alpha1.Output) Builder {
+	b.Agent.Spec.ElasticsearchRefs = refs
+	return b
+}
+
+func (b Builder) WithConfig(config *commonv1.Config) Builder {
+	b.Agent.Spec.Config = config
+	return b
+}
+
+func (b Builder) WithImage(image string) Builder {
+	b.Agent.Spec.Image = image
+	return b
+}
+
+func (b Builder) WithSuffix(suffix string) Builder {
+	if suffix != "" {
+		b.Agent.ObjectMeta.Name = b.Agent.ObjectMeta.Name + "-" + suffix
+	}
+	return b
+}
+
+func (b Builder) WithNamespace(namespace string) Builder {
+	b.Agent.ObjectMeta.Namespace = namespace
+	return b
+}
+
+func (b Builder) WithRestrictedSecurityContext() Builder {
+	b.PodTemplate.Spec.SecurityContext = test.DefaultSecurityContext()
+
+	return b
+}
+
+func (b Builder) WithContainerSecurityContext(securityContext corev1.SecurityContext) Builder {
+	for i := range b.PodTemplate.Spec.Containers {
+		b.PodTemplate.Spec.Containers[i].SecurityContext = &securityContext
+	}
+
+	return b
+}
+
+func (b Builder) WithLabel(key, value string) Builder {
+	if b.Agent.Labels == nil {
+		b.Agent.Labels = make(map[string]string)
+	}
+	b.Agent.Labels[key] = value
+
+	return b
+}
+
+func (b Builder) WithPodLabel(key, value string) Builder {
+	if b.PodTemplate.Labels == nil {
+		b.PodTemplate.Labels = make(map[string]string)
+	}
+	b.PodTemplate.Labels[key] = value
+
+	return b
+}
+
+func (b Builder) WithPodTemplateServiceAccount(name string) Builder {
+	b.PodTemplate.Spec.ServiceAccountName = name
+
+	return b
+}
+
+func (b Builder) WithRoles(clusterRoleNames ...string) Builder {
+	for _, clusterRoleName := range clusterRoleNames {
+		b = bind(b, clusterRoleName)
```

Comment:
Not sure what happens if you assign to the method receiver, my IDE does not seem to like it and I am actually unsure if the `b` in the the `bind` invocation will be propagated? Maybe introduce a variable just to avoid confusion?

Reply:
I think it warns because it thinks, that I think `b` will be replaced at the call site, while I'm just using it as a variable to carry the result. Anyway, probably better to rename 👍